Assets/Scripts/Map/Partition.cs

```diff
@@ -38,30 +38,57 @@ public sealed class Partition
         public int PartitionId { get; }
         public IReadOnlyList<Vertex> Vertices { get; }
         public Bitmap CommunicationZone { get; }
-        public IReadOnlyDictionary<Vector2Int, Bitmap> CommunicationZones { get; }
+        public IReadOnlyDictionary<Vector2Int, Bitmap> WaypointsCommunicationZones { get; }
         public IReadOnlyDictionary<int, Bitmap> IntersectionZones => _intersectionZones;
         public IReadOnlyDictionary<int, float> CommunicationRatio => _communicationRatio;
+        public List<Partition> NeighborPartitions { get; } = new();
 
         private readonly Dictionary<int, Bitmap> _intersectionZones = new();
         private readonly Dictionary<int, float> _communicationRatio = new();
+        private readonly int _bitmapWidth;
+        private readonly int _bitmapHeight;
 
         public Partition(int partitionId, IReadOnlyList<Vertex> vertices, IReadOnlyDictionary<Vector2Int, Bitmap> communicationZones)
         {
             PartitionId = partitionId;
             Vertices = vertices;
-            CommunicationZones = communicationZones;
+            WaypointsCommunicationZones = communicationZones;
             CommunicationZone = CreatePartitionCommunicationZone(vertices, communicationZones);
+            _bitmapWidth = CommunicationZone.Width;
+            _bitmapHeight = CommunicationZone.Height;
+        }
+
+        public void AddNeighborPartition(Partition partition)
+        {
+            if (NeighborPartitions.Contains(partition))
+            {
+                return;
+            }
+            NeighborPartitions.Add(partition);
+            CalculateIntersectionAndRatio(partition);
+            partition.AddNeighborPartition(this);
         }
 
         public void CalculateIntersectionAndRatio(Partition otherPartition)
         {
-            var intersection = Bitmap.Intersection(CommunicationZone, otherPartition.CommunicationZone);
-            _intersectionZones[otherPartition.PartitionId] = intersection;
+            var communicationZoneIntersection = new Bitmap(0, 0, _bitmapWidth, _bitmapHeight);
 
-            var ratio = otherPartition.CommunicationZone.Count > 0 ? (float)intersection.Count / otherPartition.CommunicationZone.Count : 0f;
+            foreach ((var position, var vertexComZone) in otherPartition.WaypointsCommunicationZones)
+            {
+                var intersection = Bitmap.Intersection(CommunicationZone, vertexComZone);
+                if (intersection.Contains(position))
+                {
+                    communicationZoneIntersection.Union(intersection);
+                }
+            }
```

Comment:
Why are we not just intersecting once with otherPartition.CommunicationZone?

Reply:
Isnt CommunicationZone the union of all the partition's waypoints' communication zones?

Reply:
We have some constraints of what we call intersecting communication zone. The intersection needs to share a waypoint such that is is actually possible to be in the intersection zone due to pathing.